File: features.py
"""
Turns raw MediaPipe results into a friendly `Features` object that gesture
detectors can read without touching MediaPipe internals.

Everything is in normalized image coordinates (0..1), so thresholds are
resolution-independent.
"""

from dataclasses import dataclass, field
from math import hypot


@dataclass
class Hand:
    """One detected hand: 21 landmarks as (x, y, z) tuples, plus handedness."""
    landmarks: list          # list[(x, y, z)]
    label: str               # "Left" or "Right" (as seen by the camera)

    # MediaPipe hand landmark indices (see the hand model diagram):
    WRIST = 0
    THUMB_TIP = 4
    INDEX_MCP = 5
    INDEX_TIP = 8
    MIDDLE_MCP = 9
    MIDDLE_TIP = 12
    RING_TIP = 16
    PINKY_TIP = 20

    def point(self, idx):
        return self.landmarks[idx]

    @property
    def index_tip(self):
        return self.landmarks[self.INDEX_TIP]


@dataclass
class Features:
    """All the per-frame data a gesture detector might need."""
    blendshapes: dict = field(default_factory=dict)  # name -> score 0..1
    face: list = field(default_factory=list)          # list[(x, y, z)] or []
    hands: list = field(default_factory=list)         # list[Hand]

    @property
    def has_face(self):
        return bool(self.face)

    @property
    def num_hands(self):
        return len(self.hands)

    def blend(self, name, default=0.0):
        """Blendshape score by name, e.g. blend('jawOpen')."""
        return self.blendshapes.get(name, default)

    def face_point(self, idx):
        return self.face[idx]

    @property
    def face_width(self):
        """Rough normalized face width (cheek to cheek). Used to scale
        distance thresholds so they don't depend on how close you sit."""
        if not self.face:
            return 0.15
        left = self.face[234]   # left cheek
        right = self.face[454]  # right cheek
        return max(abs(left[0] - right[0]), 1e-3)


def dist(a, b):
    """2D Euclidean distance between two (x, y, ...) points."""
    return hypot(a[0] - b[0], a[1] - b[1])


def build_features(face_result, hand_result):
    """Assemble a Features object from MediaPipe Tasks results."""
    f = Features()

    if face_result and face_result.face_landmarks:
        lm = face_result.face_landmarks[0]
        f.face = [(p.x, p.y, p.z) for p in lm]

    if face_result and face_result.face_blendshapes:
        for cat in face_result.face_blendshapes[0]:
            f.blendshapes[cat.category_name] = cat.score

    if hand_result and hand_result.hand_landmarks:
        handedness = hand_result.handedness or []
        for i, lm in enumerate(hand_result.hand_landmarks):
            label = "Right"
            if i < len(handedness) and handedness[i]:
                label = handedness[i][0].category_name
            f.hands.append(Hand(
                landmarks=[(p.x, p.y, p.z) for p in lm],
                label=label,
            ))

    return f

File: web/static/js/features.js
// Port of features.py — turns raw MediaPipe results into a friendly Features
// object that the gesture code reads. All coordinates are normalized (0..1),
// so thresholds are resolution-independent.

// MediaPipe hand landmark indices (see the hand model diagram).
export const HandIdx = {
  WRIST: 0,
  THUMB_TIP: 4,
  INDEX_MCP: 5,
  INDEX_TIP: 8,
  MIDDLE_MCP: 9,
  MIDDLE_TIP: 12,
  RING_TIP: 16,
  PINKY_TIP: 20,
};

export function dist(a, b) {
  return Math.hypot(a[0] - b[0], a[1] - b[1]);
}

class Hand {
  constructor(landmarks, label) {
    this.landmarks = landmarks; // [[x, y, z], ...]
    this.label = label; // "Left" | "Right"
    this.WRIST = HandIdx.WRIST;
  }
  point(idx) {
    return this.landmarks[idx];
  }
  get index_tip() {
    return this.landmarks[HandIdx.INDEX_TIP];
  }
}

class Features {
  constructor() {
    this.blendshapes = {}; // name -> score
    this.face = []; // [[x, y, z], ...] or []
    this.hands = []; // Hand[]
  }
  get has_face() {
    return this.face.length > 0;
  }
  get num_hands() {
    return this.hands.length;
  }
  blend(name, def = 0.0) {
    const v = this.blendshapes[name];
    return v === undefined ? def : v;
  }
  face_point(idx) {
    return this.face[idx];
  }
  // Rough normalized face width (cheek to cheek), used to scale distances so
  // they don't depend on how close the user sits to the camera.
  get face_width() {
    if (this.face.length === 0) return 0.15;
    const left = this.face[234];
    const right = this.face[454];
    return Math.max(Math.abs(left[0] - right[0]), 1e-3);
  }
}

// Assemble a Features object from MediaPipe Tasks results (mirrors
// build_features in features.py).
export function buildFeatures(faceResult, handResult) {
  const f = new Features();

  const faceLandmarks = faceResult?.faceLandmarks;
  if (faceLandmarks && faceLandmarks.length) {
    f.face = faceLandmarks[0].map((p) => [p.x, p.y, p.z]);
  }

  const blendshapes = faceResult?.faceBlendshapes;
  if (blendshapes && blendshapes.length) {
    for (const cat of blendshapes[0].categories) {
      f.blendshapes[cat.categoryName] = cat.score;
    }
  }

  const handLandmarks = handResult?.landmarks;
  if (handLandmarks && handLandmarks.length) {
    // The plural key changed across MediaPipe versions; accept either.
    const handedness = handResult.handednesses || handResult.handedness || [];
    handLandmarks.forEach((lm, i) => {
      let label = "Right";
      if (handedness[i] && handedness[i][0]) {
        label = handedness[i][0].categoryName;
      }
      f.hands.push(new Hand(lm.map((p) => [p.x, p.y, p.z]), label));
    });
  }

  return f;
}

export { Features, Hand };
